### PR TITLE
rockchip: add basic support for Bananapi-R2 Pro

### PR DESCRIPTION
**Pull request by antnyfls**
Tested working:
- USB 3.0 ports x2
- SATA port
- Mini PCIe and M.2-E interfaces
- MT7531BE Swich x4 LAN ports
- RTL8211F WAN port
- SD Card and eMMC
- Power and Reset buttons
- CH340 micro USB debug UART port
- FAN connector
- RTC battery

Not working: 

- HDMI

Untested

- SIM card
- Speaker connector
- Headphones
- Mic
- IR
- MIPI connectors

[Boot Logs](https://gist.github.com/antnyfls/bb716aa230a58350a8fb4038691803cc)

To boot from the SD card
1-hold down the MaskRom button
2-Connect DC power
3-Wait 5 seconds, release the button.

### CODE REVIEW

File: target/linux/rockchip/armv8/base-files/etc/board.d/01_leds [hunk 1]
```diff
@@ -28,6 +28,10 @@ friendlyarm,nanopi-r5s)
 	ucidef_set_led_netdev "lan1" "LAN1" "green:lan-1" "eth1"
 	ucidef_set_led_netdev "lan2" "LAN2" "green:lan-2" "eth2"
 	;;
+sinovoip,rk3568-bpi-r2pro)
+	ucidef_set_led_netdev "wan" "WAN" "blue:status" "eth0"
```
Comment (1715173329) [inline target/linux/rockchip/armv8/base-files/etc/board.d/01_leds:32]: this is not okay... the LED on-board is not used for network activity.

File: target/linux/rockchip/armv8/base-files/etc/board.d/02_network [hunk 2]
```diff
@@ -56,6 +59,10 @@ rockchip_setup_macs()
 		wan_mac=$(macaddr_generate_from_mmc_cid mmcblk*)
 		lan_mac=$(macaddr_add "$wan_mac" 1)
 		;;
+	sinovoip,rk3568-bpi-r2pro)
```
Comment (1715173329) [inline target/linux/rockchip/armv8/base-files/etc/board.d/02_network:62]: no, this is duplicated. and this board has emmc module so we should generate macaddr from there.
```
	friendlyarm,nanopi-r2c-plus|\
	friendlyarm,nanopi-r4s|\
	friendlyarm,nanopi-r5s|\
	sinovoip,rk3568-bpi-r2pro)
```

File: target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity [hunk 3]
```diff
@@ -53,5 +53,9 @@ friendlyarm,nanopi-r5s)
 	set_interface_core 4 "eth1"
 	set_interface_core 8 "eth2"
 	;;
+sinovoip,rk3568-bpi-r2pro)
```
Comment (1715173329) [inline target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity:56]: duplicated.

```
friendlyarm,nanopi-r5c|\
radxa,e25|\
sinovoip,rk3568-bpi-r2pro)
```

File: target/linux/rockchip/armv8/config-6.1 [hunk 4]
```diff
@@ -363,6 +363,7 @@ CONFIG_MDIO_BUS_MUX_GPIO=y
 CONFIG_MDIO_BUS_MUX_MMIOREG=y
 CONFIG_MDIO_DEVICE=y
 CONFIG_MDIO_DEVRES=y
+CONFIG_MEDIATEK_GE_PHY=y
```
Comment (1715173329) [inline target/linux/rockchip/armv8/config-6.1:366]: why? this is only used for mediatek's built-in Ge PHY.

File: target/linux/rockchip/image/armv8.mk [hunk 5]
```diff
@@ -130,6 +130,15 @@ define Device/radxa_rock-pi-e
 endef
 TARGET_DEVICES += radxa_rock-pi-e
 
+define Device/rockchip_bpi-r2-pro
```
Comment (1715173329) [inline target/linux/rockchip/image/armv8.mk:133]:
```suggestion
define Device/sinovoip_bpi-r2-pro
```

**Review by 1715173329 — CHANGES_REQUESTED** [review]

File: target/linux/rockchip/armv8/config-6.1 (hunk 4 above)
Comment (musashino205) [reply target/linux/rockchip/armv8/config-6.1:366]: `CONFIG_MEDIATEK_GE_SOC_PHY` is for built-in PHYs in MediaTek MT798x SoCs, but `CONFIG_MEDIATEK_GE_PHY` is for the PHYs in MediaTek MT753x switches. This is selected by `CONFIG_NET_DSA_MT7530_MDIO`.

https://elixir.bootlin.com/linux/v6.1.85/source/drivers/net/phy/mediatek-ge.c#L71

File: target/linux/rockchip/armv8/base-files/etc/board.d/01_leds (hunk 1 above)
Comment (antnyfls) [reply target/linux/rockchip/armv8/base-files/etc/board.d/01_leds:32]: Okay, I removed it

File: target/linux/rockchip/armv8/base-files/etc/board.d/02_network (hunk 2 above)
Comment (antnyfls) [reply target/linux/rockchip/armv8/base-files/etc/board.d/02_network:62]: The board can boot from  eMMC  and SD card, so the correct is _mmcblk*_
```
friendlyarm,nanopi-r5c|\
sinovoip,rk3568-bpi-r2pro)
		wan_mac=$(macaddr_generate_from_mmc_cid mmcblk*)
		lan_mac=$(macaddr_add "$wan_mac" 1)
```

File: target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity (hunk 3 above)
Comment (antnyfls) [reply target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity:56]: done

File: target/linux/rockchip/image/armv8.mk (hunk 5 above)
Comment (antnyfls) [reply target/linux/rockchip/image/armv8.mk:133]: done

File: target/linux/rockchip/armv8/base-files/etc/board.d/02_network (hunk 2 above)
Comment (1715173329) [reply target/linux/rockchip/armv8/base-files/etc/board.d/02_network:59]: > The board can boot from eMMC and SD card, so the correct is _mmcblk*_

Current practice is to get a "fixed" mac address. The generated macaddr will be different if user replaces SD card, compared of that macaddr generated using eMMC module is more "stable".

**Comment by TahomaSoft** [comment]
I was working on this board independently and just put in pull request 15162 to also support BPI-R2 Pro before I saw this recent PR.  Happy to collaborate together.

**Comment by 1715173329** [comment]
For the second commit title, would be better to use `rockchip: enable MT7531 and RTC drivers`.

I still suggest using `mmcblk1` for macaddr generation.
I use `mmcblk*` for nanopi r5c because on that board the emmc module is optional, but on bpi-r2 pro it isn't.

Otherwise:
`Reviewed-by: Tianling Shen <cnsztl@immortalwrt.org>`

Thanks.

**Comment by antnyfls** [comment]
@1715173329 if I follow your your suggestion (mmcblk1) the device will generate a mac address only from the eMMC, now in the case, the user want to use a SD card (mmcblk0) the script will not work.

**Comment by 1715173329** [comment]
> the device will generate a mac address only from the eMMC

https://github.com/openwrt/openwrt/pull/15137#discussion_r1563587353
It's expected ;)

**Comment by TahomaSoft** [comment]
Use of the SD card is more important than utilizing the eMMC to me.

**Comment by musashino205** [comment]
IMO, a device's MAC address should be as unique and permanent as possible, so it makes sense to use eMMC for generation.

**Comment by antnyfls** [comment]
> IMO, a device's MAC address should be as unique and permanent as possible, so it makes sense to use eMMC for generation.

@musashino205 Obviously the eMMC should be the preferred method for the mac address generations, issue come, on this device, when boot from a SD card the eMMC is not detected by the script.

**Comment by TahomaSoft** [comment]
The device will generate a MAC address if the script isn't used; my work on this (before I found @antnyfls' pull request) didn't use a MAC generating script, but the device did assign itself one. I'm not sure how stable that self-assigned address was across boots. I could check.

**Comment by 1715173329** [comment]
> on this device, when boot from a SD card the eMMC is not detected by the script.

Hmmm what do you mean by "not detected"? Does the eMMC device disappear from `/dev`?

**Comment by antnyfls** [comment]
> Hmmm what do you mean by "not detected"? Does the eMMC device disappear from `/dev`?

@1715173329 yes not detected, it appears it was related to https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/arch/arm64/boot/dts/rockchip/rk3568-bpi-r2-pro.dts?h=next-20240423&id=433d54818f64a2fe0562f8c04c7a81f562368515 I've backported the fix  from upstream linux next.

**Comment by TahomaSoft** [comment]
Hi, so is this PR ready for merging?

**Comment by 1715173329** [comment]
Yes. Waiting a maintainer to pick it.

**Comment by TahomaSoft** [comment]
Great, thanks!  Hopefully the @CODEOWNERS will pick this up and approve it soon.

**Review by TahomaSoft — APPROVED** [review]
Looks lovely to me.

**Comment by TahomaSoft** [comment]
It might be worth mentioning here to expedite merging that this PR builds on support already upstream in the canonical u-boot tree and kernel tree. Those repositories already support this board, the BPI-R2-Pro.

**Comment by 1715173329** [comment]
Cc: @hauke @PolynomialDivision
Any chance to take a look? Thanks!

**Comment by TahomaSoft** [comment]
Any CODEOWNERS available to take a look?

**Comment by TahomaSoft** [comment]
Thank you!!!!!!
:+)

**Pull request merged**